File: src/main/java/pl/mroz/buddiesapi/domain/rental/RentalRepository.java
package pl.mroz.buddiesapi.domain.rental;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.domain.Specification;
import pl.mroz.buddiesapi.domain.common.Location;
import pl.mroz.buddiesapi.infrastructure.database.account.AccountEntity;
import pl.mroz.buddiesapi.infrastructure.database.location.LocationEntity;
import pl.mroz.buddiesapi.infrastructure.database.rental.RentalEntity;

import java.time.Instant;
import java.util.List;
import java.util.Set;
import java.util.UUID;

public interface RentalRepository {

    List<Rental> getAllRentals();

    Page<Rental> getByCriteria(Specification<RentalEntity> criteria, Pageable pageable);

    Rental getRentalById(UUID rentalId);

    List<Rental> getAllRentalsFromUser(UUID accountId);

    List<String> getTopTags();

    Rental save(Rental rental);

    void delete(Rental rental);

    interface IRentalEntity {

        UUID getRentalId();

        AccountEntity getAuthorEntity();

        String getTitle();

        boolean isNegotiable();

        String getDescription();

        LocationEntity getLocationEntity();

        int getPrice();

        int getDeposit();

        int getRooms();

        int getFloor();

        double getSize();

//        double getPricePerM();

        int getBuildYear();

        Instant getRentDate();

//        Set<String> getFeatureTags();

//        Set<String> getPhotoUrls();

    }

    interface IRentalCriteria {

        int getMaxDistance();

        Location getLocation();

        int getPriceMin();

        int getPriceMax();

        int getSizeMin();

        int getSizeMax();

        int getRoomsMin();

        int getRoomsMax();

        int getBuildMin();

        int getBuildMax();

        int getMoveMin();

        int getMoveMax();

        Set<String> getFeatures();
    }
}

File: src/main/kotlin/buddiesapi/domain/rental/RentalRepository.kt
package buddiesapi.domain.rental

import buddiesapi.domain.common.Location
import buddiesapi.infrastructure.database.account.AccountEntity
import buddiesapi.infrastructure.database.location.LocationEntity
import buddiesapi.infrastructure.database.rental.RentalEntity
import java.time.Instant
import java.util.UUID
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.jpa.domain.Specification

interface RentalRepository {
    fun getAllRentals(): List<Rental>

    fun getByCriteria(criteria: Specification<RentalEntity>, pageable: Pageable): Page<Rental>

    fun getRentalById(rentalId: UUID): Rental?

    fun getAllRentalsFromUser(accountId: UUID): List<Rental>

    val topTags: List<String>

    fun save(rental: Rental): Rental

    fun delete(rental: Rental)

    interface IRentalEntity {
        val rentalId: UUID?

        val authorEntity: AccountEntity

        val title: String

        val isNegotiable: Boolean

        val description: String

        val locationEntity: LocationEntity

        val price: Int

        val deposit: Int

        val rooms: Int

        val floor: Int

        val size: Double

        //        double getPricePerM();
        val buildYear: Int

        //        Set<String> getFeatureTags();
        val rentDate: Instant

        //        Set<String> getPhotoUrls();
    }

    interface IRentalCriteria {
        val maxDistance: Int?

        val location: Location?

        val priceMin: Int?

        val priceMax: Int?

        val sizeMin: Int?

        val sizeMax: Int?

        val roomsMin: Int?

        val roomsMax: Int?

        val buildMin: Int?

        val buildMax: Int?

        val moveMin: Int?

        val moveMax: Int?

        val features: Set<String?>?
    }
}
